File: leetcode 1326.cpp
class Solution {
public:
  int minTaps(int n, vector<int> &ranges) {
    std::vector<int> right_bound(n + 1, 0);
    for (int i = 0; i < n + 1; i++) {
      int left = std::max(0, i - ranges[i]);
      right_bound[left] = std::max(right_bound[left], i + ranges[i]);
    }
    int cur_right = 0, next_right = 0, ans = 0;
    for (int i = 0; i < n; i++) {
      next_right = std::max(next_right, right_bound[i]);
      if (i == cur_right) {
        if (next_right < i + 1) {
          return -1;
        }
        cur_right = next_right;
        ans++;
      }
    }
    return ans;
  }
};

File: leetcode 1326.py
from typing import List

class Solution:
    def minTaps(self, n: int, ranges: List[int]) -> int:
        right_bound = [0] * (n + 1) # 每个位置能延伸的右边界
        for i, rangei in enumerate(ranges):
            li = max(0, i - rangei)
            right_bound[li] = max(right_bound[li], i + rangei)
        cur_right = 0 # 当前已经存在的桥
        next_right = 0 # 下一座要修的桥
        ans = 0
        # 不需要遍历到最后一个位置
        for i in range(n):
            next_right = max(next_right, right_bound[i])
            # 到了当前桥的最远距离
            if i == cur_right:
                if next_right < i + 1:
                    # 到不了下一个位置了
                    return -1
                cur_right = next_right
                ans += 1
        return ans